# 4.8 Task: Grouping and Aggregating

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

This script contains the following points:
1. Importing orders and products data
2. Grouping and aggregrating values
3. Creating user-defined field for Loyalty_flag, spending_flag, and frequency_flag

In [2]:
path=r'C:\Users\dell\Desktop\Career Foundry\0423_Instacart Analysis'

In [3]:
# Import pickle for order_products_combined data set
ords_prods_merge=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ord_prod_newvar.pkl'))

In [4]:
# Group and mean of department_id

ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Analyzing the mean for the entire data set against the sub_set, additional department_ids now appear. Most of the averages decreased from the values in the subset with the exception of department_id 17; which increased from ~11.29 to ~15.69.

In [5]:
# Creating max_order field via max aggregate for Loyalty Flag

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

# Setting Loyalty Flags

In [6]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [7]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [8]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [9]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [10]:
ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   #ofdays_since_last_order  product_id  add_to_cart_order  reordered  \
0                       0.0         196                  1          0   
1                      15.0         196                  1          1   
2                      21.0         196                  1          1   
3                      29.0         196                  1          1   
4                      28.0         196                  1          1   
5                      19.0         196                  1          1   
6                      20.0         196                  1          1   
7                      14.0         196                  2          1   
8                       0.0         196                  4          1   
9                      30.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

   busiest_days busiest_period_of_day  max_order  loyalty_flag  
0  Average Days        Average orders         10  New customer  
1  Slowest Days        Average orders         10  New customer  
2  Slowest Days        Average orders         10  New customer  
3  Slowest Days        Average orders         10  New customer  
4  Slowest Days        Average orders         10  New customer  
5  Average Days        Average orders         10  New customer  
6  Busiest Days        Average orders         10  New customer  
7  Busiest Days        Average orders         10  New customer  
8  Busiest Days        Average orders         10  New customer  
9  Slowest Days        Average orders         10  New customer

In [11]:
# Grouping loyalty_flag with mean, min, max prices

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

In [12]:
# Creating mean_spend field via mean aggregate for Spending Flag

ords_prods_merge['mean_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [13]:
ords_prods_merge[['user_id','mean_spend','order_number']].head(15)

user_id  mean_spend  order_number
0         1    6.367797             1
1         1    6.367797             2
2         1    6.367797             3
3         1    6.367797             4
4         1    6.367797             5
5         1    6.367797             6
6         1    6.367797             7
7         1    6.367797             8
8         1    6.367797             9
9         1    6.367797            10
10       15    3.980556            15
11       15    3.980556            17
12       15    3.980556            18
13       15    3.980556            21
14       15    3.980556            22

# Setting Spending Flags

In [14]:
ords_prods_merge.loc[ords_prods_merge['mean_spend'] < 10, 'spending_flag'] = 'Low spender'

In [15]:
ords_prods_merge.loc[ords_prods_merge['mean_spend'] >= 10, 'spending_flag'] = 'High spender'

In [16]:
ords_prods_merge['spending_flag'].value_counts(dropna=False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [17]:
# Creating median_spend field via mean aggregate for Frequency Flag

ords_prods_merge['median_freq'] = ords_prods_merge.groupby(['user_id'])['#ofdays_since_last_order'].transform(np.median)

In [18]:
ords_prods_merge[['user_id','median_freq','
                  ']].head(15)

user_id  median_freq  #ofdays_since_last_order
0         1         20.0                       0.0
1         1         20.0                      15.0
2         1         20.0                      21.0
3         1         20.0                      29.0
4         1         20.0                      28.0
5         1         20.0                      19.0
6         1         20.0                      20.0
7         1         20.0                      14.0
8         1         20.0                       0.0
9         1         20.0                      30.0
10       15          8.0                       7.0
11       15          8.0                       8.0
12       15          8.0                       7.0
13       15          8.0                       7.0
14       15          8.0                      14.0

User_id returns a median_freq of 8, but median appears to be 7

# Setting Frequency Flags

In [19]:
ords_prods_merge.loc[ords_prods_merge['median_freq'] > 20, 'frequency_flag'] = 'Non-Frequent customer'

In [20]:
ords_prods_merge.loc[(ords_prods_merge['median_freq'] <= 20) & (ords_prods_merge['median_freq'] > 10), 'frequency_flag'] = 'Regular customer'

In [21]:
ords_prods_merge.loc[ords_prods_merge['median_freq'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [22]:
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent customer        22796659
Regular customer          6921472
Non-Frequent customer     2686728
Name: frequency_flag, dtype: int64

In [23]:
ords_prods_merge[['user_id','median_freq','frequency_flag']].head(15)

user_id  median_freq     frequency_flag
0         1         20.0   Regular customer
1         1         20.0   Regular customer
2         1         20.0   Regular customer
3         1         20.0   Regular customer
4         1         20.0   Regular customer
5         1         20.0   Regular customer
6         1         20.0   Regular customer
7         1         20.0   Regular customer
8         1         20.0   Regular customer
9         1         20.0   Regular customer
10       15          8.0  Frequent customer
11       15          8.0  Frequent customer
12       15          8.0  Frequent customer
13       15          8.0  Frequent customer
14       15          8.0  Frequent customer

In [24]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   #ofdays_since_last_order  product_id  add_to_cart_order  reordered  \
0                       0.0         196                  1          0   
1                      15.0         196                  1          1   
2                      21.0         196                  1          1   
3                      29.0         196                  1          1   
4                      28.0         196                  1          1   

  product_name  ...  prices    price_range_loc  busiest_days  \
0         Soda  ...     9.0  Mid-range product  Average Days   
1         Soda  ...     9.0  Mid-range product  Slowest Days   
2         Soda  ...     9.0  Mid-range product  Slowest Days   
3         Soda  ...     9.0  Mid-range product  Slowest Days   
4         Soda  ...     9.0  Mid-range product  Slowest Days   

  busiest_period_of_day max_order  loyalty_flag  mean_spend spending_flag  \
0        Average orders        10  New customer    6.367797   Low spender   
1        Average orders        10  New customer    6.367797   Low spender   
2        Average orders        10  New customer    6.367797   Low spender   
3        Average orders        10  New customer    6.367797   Low spender   
4        Average orders        10  New customer    6.367797   Low spender   

   median_freq    frequency_flag  
0         20.0  Regular customer  
1         20.0  Regular customer  
2         20.0  Regular customer  
3         20.0  Regular customer  
4         20.0  Regular customer  

[5 rows x 22 columns]

In [25]:
# Export data to pkl

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_agg.pkl'))